File: puzzles/ts/dwarfs-giants/dwarfs-giants.ts
function topologicalSort(graph, nodeId: number, visited, stack: number[]) {
  visited.add(nodeId);

  for (let child of graph.get(nodeId)) {
    if (!visited.has(child)) {
      topologicalSort(graph, child, visited, stack);
    }
  }

  stack.push(nodeId);
}

function longestPath(graph): number {
  let stack: number[] = [];
  let distances = new Map<number, number>(); // node id => max distance to this node
  let visited = new Set<number>();

  for (let key of graph.keys()) {
    if (!visited.has(key)) {
      topologicalSort(graph, key, visited, stack);
    }
  }

  for (let key of graph.keys()) {
    distances.set(key, 1);
  }

  while (stack.length != 0) {
    const nodeId: number = stack.pop();
    let currentDistance = distances.get(nodeId);
    for (let child of graph.get(nodeId)) {
      if (distances.get(child) < currentDistance + 1) {
        distances.set(child, currentDistance + 1);
      }
    }
  }

  let maxDistance: number = 0;
  for (let key of distances.keys()) {
    maxDistance = Math.max(maxDistance, distances.get(key));
  }
  return maxDistance;
}

let graph = new Map<number, number[]>(); // node id => array of children ids
const n: number = Number(readline()); // the number of relationships of influence

for (let i: number = 0; i < n; i++) {
  const inputs: string[] = readline().split(' ');
  const x: number = Number(inputs[0]); // a relationship of influence between two people (x influences y)
  const y: number = Number(inputs[1]);
  if (!graph.has(x)) {
    graph.set(x, []);
  }
  if (!graph.has(y)) {
    graph.set(y, []);
  }
  graph.get(x).push(y);
}

// The number of people involved in the longest succession of influences
console.log(longestPath(graph));

File: puzzles/python3/dwarfs-giants/dwarfs_giants.py
from typing import Dict, Set


class Node:
    def __init__(self):
        self.children: Set[int] = set()
        self.depth = 0


class Graph:
    def __init__(self):
        self.nodes: Dict[int, Node] = {}  # node_id => node

    def build(self):
        nb_edges = int(input())  # the number of relationships of influence
        for _ in range(nb_edges):
            parent_id, child_id = map(int, input().split())

            if child_id in self.nodes:
                child: Node = self.nodes[child_id]
            else:
                child: Node = Node()
                self.nodes[child_id] = child

            if parent_id in self.nodes:
                parent: Node = self.nodes[parent_id]
                parent.children.add(child_id)
            else:
                parent: Node = Node()
                parent.children.add(child_id)
                self.nodes[parent_id] = parent

    def solve(self) -> int:
        """Returns the number of people involved in the longest succession of influences."""
        max_depth = 0
        for node in self.nodes.values():
            self.traverse(node)
        for node in self.nodes.values():
            if node.depth > max_depth:
                max_depth = node.depth
        return max_depth + 1

    def traverse(self, node: Node):
        for child_id in node.children:
            child: Node = self.nodes[child_id]
            if child.depth <= node.depth:
                child.depth = node.depth + 1
                self.traverse(child)


if __name__ == "__main__":
    graph = Graph()
    graph.build()
    print(graph.solve())
